protobuf/plugin/customname.go

```diff
@@ -0,0 +1,50 @@
+package plugin
+
+import (
+	"strings"
+
+	"github.com/gogo/protobuf/gogoproto"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/gogo/protobuf/protoc-gen-gogo/generator"
+)
+
+// CustomName preprocess the field, and set the [(gogoproto.customname) = "..."]
+// if necessary, in order to avoid setting `gogoproto.customname` manually.
+// The automatically assigned name should conform to Golang convention.
+func CustomName(field *descriptor.FieldDescriptorProto) {
+	// Skip if [(gogoproto.customname) = "..."] has already been set.
+	if gogoproto.IsCustomName(field) {
+		return
+	}
+	// Skip if embedded
+	if gogoproto.IsEmbed(field) {
+		return
+	}
+	if field.OneofIndex != nil {
+		return
+	}
+	fieldName := generator.CamelCase(*field.Name)
+	switch {
+	case *field.Name == "id":
+		// id -> ID
+		fieldName = "ID"
+	case strings.HasPrefix(*field.Name, "id_"):
+		// id_some -> IDSome
+		fieldName = "ID" + fieldName[2:]
+	case strings.HasSuffix(*field.Name, "_id"):
+		// some_id -> SomeID
+		fieldName = fieldName[:len(fieldName)-2] + "ID"
+	case strings.HasSuffix(*field.Name, "_ids"):
+		// some_ids -> SomeIDs
+		fieldName = fieldName[:len(fieldName)-3] + "IDs"
```

Comment:
@stevvooe I didn't find the usage of `_id_` in current `.proto` files. Here is the remaining `(gogoproto.customname)`:
```
ubuntu@ubuntu:~/swarmkit$ grep -r gogoproto.customname api
api/objects.proto:	repeated VirtualIP virtual_ips = 3 [(gogoproto.customname) = "VirtualIPs"];
api/objects.proto:	IPAMOptions ipam = 5 [(gogoproto.customname) = "IPAM"];
api/objects.proto:	RootCA root_ca = 4 [(gogoproto.nullable)=false, (gogoproto.customname) = "RootCA"];
api/types.proto:	int64 nano_cpus = 1 [(gogoproto.customname) = "NanoCPUs"];
api/types.proto:	string os = 2 [(gogoproto.customname) = "OS"];
api/types.proto:	bool readonly = 4 [(gogoproto.customname) = "ReadOnly"];
api/types.proto:		bool nocopy = 1 [(gogoproto.customname) = "NoCopy"];
api/types.proto:	int32 pid = 2 [(gogoproto.customname) = "PID"];
api/types.proto:	string url = 2 [(gogoproto.customname) = "URL"];
api/types.proto:	repeated ExternalCA external_cas = 2 [(gogoproto.customname) = "ExternalCAs"];
api/types.proto:	bytes ca_key = 1 [(gogoproto.customname) = "CAKey"];
api/types.proto:	bytes ca_cert = 2 [(gogoproto.customname) = "CACert"];
api/types.proto:	string ca_cert_hash = 3 [(gogoproto.customname) = "CACertHash"];
api/types.proto:	bytes csr = 2 [(gogoproto.customname) = "CSR"];
api/types.proto:	string cn = 5 [(gogoproto.customname) = "CN"];
api/types.proto:		string uid = 2 [(gogoproto.customname) = "UID"];
api/types.proto:		string gid = 3 [(gogoproto.customname) = "GID"];
api/ca.proto:	bytes csr = 2 [(gogoproto.customname) = "CSR"];
api/specs.proto:	bool tty = 13 [(gogoproto.customname) = "TTY"];
api/specs.proto:	DNSConfig dns_config = 15 [(gogoproto.customname) = "DNSConfig"];
api/specs.proto:	IPAMOptions ipam = 5 [(gogoproto.customname) = "IPAM"];
api/specs.proto:	CAConfig ca_config = 6 [(gogoproto.nullable) = false, (gogoproto.customname) = "CAConfig"];
ubuntu@ubuntu:~/swarmkit$ 
```